d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -245,29 +254,40 @@ private void handleD2URIMapResponse(DiscoveryResponseData data)
       {
         XdsD2.D2URIMap uriMap = resource.getResource().unpack(XdsD2.D2URIMap.class);
         Map<String, XdsD2.D2URI> nodeData = uriMap.getUrisMap();
-        updates.put(resourceName, new D2URIMapUpdate(resource.getVersion(), nodeData));
-      } catch (InvalidProtocolBufferException e)
+        if (nodeData.isEmpty())
+        {
+          _log.warn("Received a D2URIMap response with no data, resource is : {}", resourceName);
+        }
+        else
+        {
+          updates.put(resourceName, new D2URIMapUpdate(resource.getVersion(), nodeData));
+        }
+      }
+      catch (InvalidProtocolBufferException e)
       {
         _log.warn("Failed to unpack D2URIMap response", e);
         errors.add("Failed to unpack D2URIMap response");
       }
     }
-
-    handleResourceUpdate(updates, data.getResourceType(), data.getNonce(), errors);
+    sendAckOrNack(data.getResourceType(), data.getNonce(), errors);
+    handleResourceUpdate(updates, data.getResourceType());
```

Comment:
also need to call handleResourceRemoval, doesn't it?

Reply:
Yeah. Do u mean `handleResourceRemoval(data.getRemovedResources(), data.getResourceType());` in next line ? I saw the change code include this ?